### ISSUE TITLE
[graph_trainer] dynamic chunking blocked by rank-3 biasless linear with mark_unbacked

### ISSUE THREAD

**Issue by sanketpurandare** [pytorch/torchtitan#3322]
# GraphTrainer dynamic chunking: `mark_unbacked` batch dim fails on biasless rank-3 `linear`

## Summary

GraphTrainer chunk passes need to mark selected activation dimensions dynamic before tracing so the chunked graph can split batch/sequence dimensions symbolically. This currently works for MoE-only regions, but tracing an attention-containing TransformerBlock with dynamic batch fails before the chunk pass runs.

The failure is caused by biasless rank-3 `nn.Linear` / `F.linear` on an input with an unbacked symbolic leading dimension. This is hit by DeepSeekV3 attention projections such as `self.wq(x)`, where `x` is `[batch, seq, hidden]` and the projection has `bias=False`.

## Minimal Repro

Run from a TorchTitan checkout:

```bash
 python /tmp/repro_graph_trainer_unbacked_rank3_linear.py
```

Script:

```python
import traceback

import torch
import torch.nn.functional as F
from torch._dynamo.decorators import mark_unbacked

from torchtitan.experiments.graph_trainer.make_fx_tracer import (
    minimal_fx_tracer,
    run_traced,
)


def trace_case(name, fn, x_shape):
    print(f"\n{name}")
    x = torch.randn(*x_shape, device="cuda")
    w = torch.randn(3072, 256, device="cuda")
    mark_unbacked(x, 0, min=2, max=8, hint_override=x_shape[0])

    try:
        traced = minimal_fx_tracer(lambda state, x: fn(x, state["w"]))(
            {"w": w},
            x,
        )
        replay_x = torch.randn(5, *x_shape[1:], device="cuda")
        y = run_traced(traced, {"w": w}, replay_x)
        print("OK", tuple(y.shape))
    except Exception as exc:
        print(type(exc).__name__)
        print(exc)
        traceback.print_exc(limit=8)


def linear_2d(x, w):
    return F.linear(x, w)


def linear_3d_biasless(x, w):
    return F.linear(x, w)


def linear_3d_explicit_flatten(x, w):
    bsz = x.size(0)
    seqlen = x.size(1)
    y = F.linear(x.reshape(-1, x.size(-1)), w)
    return y.reshape(bsz, seqlen, -1)


if __name__ == "__main__":
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required for this repro")

    trace_case("2D biasless linear: expected OK", linear_2d, (4, 256))
    trace_case("3D biasless linear: currently fails", linear_3d_biasless, (4, 16, 256))
    trace_case(
        "3D explicit flatten + linear + reshape: expected OK",
        linear_3d_explicit_flatten,
        (4, 16, 256),
    )
```

Observed output:

```text
2D biasless linear: expected OK
OK (5, 3072)

3D biasless linear: currently fails
RuntimeError
Cannot call numel() on tensor with symbolic sizes/strides

3D explicit flatten + linear + reshape: expected OK
OK (5, 16, 3072)
```

## Why This Matters

GraphTrainer chunk passes need to trace modules with symbolic batch/sequence dimensions. For DSv3 MoE-only regions, this works and we can validate bitwise equivalence against eager chunking. For a full TransformerBlock, tracing fails at the first attention projection before chunking has a chance to run.

The relevant model path is:

```text
DeepSeekV3TransformerBlock.forward
-> Attention.forward
-> q = self.wq(x)
-> biasless rank-3 F.linear(x, weight)
-> GuardOnDataDependentSymNode
```

## Root Cause Hypothesis

2D biasless linear lowers to `mm`, whose meta implementation can preserve the symbolic leading dimension.

3D biasless linear goes through the `matmul` path. The matmul folding decision checks emptiness / folding conditions using a `numel()`-derived expression. With input shape `[u0, 16, 256]`, that expression depends on unbacked symbol `u0`. Even with positive bounds from `mark_unbacked(x, 0, min=2, max=8)`, the current fake/meta path asks for a concrete/specialized integer and raises.

In the full DeepSeekV3 TransformerBlock trace this may surface as either:

```text
Cannot call numel() on tensor with symbolic sizes/strides
```

or:

```text
GuardOnDataDependentSymNode: Could not extract specialized integer from data-dependent expression u0
```

depending on the exact tracing path.

## Desired Behavior

Tracing should succeed without specializing the batch dimension to the hint. The resulting graph should keep the leading dimension symbolic and replay with another valid batch size.

## Upstream PyTorch Fix Direction

Preferred upstream direction:

> Teach the `matmul` folding emptiness check to reason from symbolic bounds without forcing a concrete integer.

Specifically, if an unbacked symbol has a positive lower bound, `numel() == 0` for shapes like `[u0, 16, 256]` should be provably false, so the folding decision should not require concretizing `u0`.
